src/surface_rl_decoder/surface_code_util.py

```diff
@@ -28,22 +33,34 @@ def check_final_state(actual_errors, actions):
     final_state: (h, d, d) array of final state,
         i.e. all proposed actions executed on initial qubit state
     is_ground_state: (bool) shows whether final_state is in the ground state or not
+    (n_syndromes, n_loops): (tuple) counting the number of syndromes and/or loops
+        remaining in the final layer after all corrections have been performed
     """
-    # check for trivial loops
-    # check for logical operation
-    # check if still errors left
 
-    final_state = perform_all_actions(actual_errors, actions)
+    final_qubit_configuration = perform_all_actions(actual_errors, actions)
+    final_state = create_syndrome_output(
+        final_qubit_configuration[-1],
+        vertex_mask=vertex_mask,
+        plaquette_mask=plaquette_mask,
+    )
+    final_state = final_state.reshape(1, final_state.shape[-2], final_state.shape[-1])
+
+    # look for uncorrected syndromes
+    if (n_syndromes := final_state[-1].astype(np.uint8).sum()) != 0:
+        return final_state, False, (n_syndromes, 0)
 
-    z_errors = (final_state[-1] == 3).astype(np.uint8)
-    y_errors = (final_state[-1] == 2).astype(np.uint8)
-    x_errors = (final_state[-1] == 1).astype(np.uint8)
+    # no syndromes left
+    # check for non-trivial loops
+    z_errors = (final_qubit_configuration[-1] == 3).astype(np.uint8)
+    y_errors = (final_qubit_configuration[-1] == 2).astype(np.uint8)
+    x_errors = (final_qubit_configuration[-1] == 1).astype(np.uint8)
 
     x_matrix = x_errors + y_errors
     z_matrix = y_errors + z_errors
 
     x_loops = np.sum(np.sum(x_matrix, axis=0))
-    z_loops = np.sum(np.sum(z_matrix, axis=0))
+    z_loops = np.sum(np.sum(z_matrix != 0, axis=0))
+    n_loops = x_loops + z_loops
```

Comment:
Right now, we're counting the absolute number of involved qubits to form a non-trivial loop.
It would probably suffice to count one non-trivial loop as one occurence